File: space-game/Entities/BaseGameEntity.vb
Imports Messaging

''' <summary>
''' Base class for all game objects
''' </summary>
Public MustInherit Class BaseGameEntity

    ' every entity has a unique identifier
    Private ReadOnly _id As Guid
    Protected ReadOnly _entityManager As IEntityManager

    Public ReadOnly Property Id As Guid
        Get
            Return _id
        End Get
    End Property


    Protected Sub New(entityManager As IEntityManager)

        If entityManager Is Nothing Then
            Throw New ArgumentNullException(NameOf(entityManager))
        End If

        _id = Guid.NewGuid()
        _entityManager = entityManager
        _entityManager.RegisterEntity(Me)
    End Sub

    ''' <summary>
    ''' All entities must implement an update function
    ''' </summary>
    ''' <remarks></remarks>
    Public MustOverride Sub Update()

    ''' <summary>
    ''' All subclasses can communicate using messages
    ''' </summary>
    Public Overridable Function HandleMessage(msg As Telegram) As Boolean
        Throw New NotImplementedException
    End Function

    ''' <summary>
    ''' Type of the entity
    ''' </summary>
    Public ReadOnly Property Type As String
        Get
            Return Me.GetType().ToString()
        End Get
    End Property

    ''' <summary>
    ''' Whether the entity can be used as a camera target
    ''' </summary>
    Public Overridable Property CameraTarget As Boolean = True

End Class

File: space-game/Entities/BaseGameEntity.cs
using Messaging;
using System;

namespace Entities
{
    /// <summary>
    /// Base class for all game objects
    /// </summary>
    public abstract class BaseGameEntity
    {
        // every entity has a unique identifier
        private readonly Guid _id;

        protected readonly IEntityManager _entityManager;
        public Guid Id => _id;

        protected BaseGameEntity(IEntityManager entityManager)
        {
            if (entityManager == null)
            {
                throw new ArgumentNullException(nameof(entityManager));
            }

            _id = Guid.NewGuid();
            _entityManager = entityManager;
            _entityManager.RegisterEntity(this);
        }

        /// <summary>
        /// All entities must implement an update function
        /// </summary>
        /// <remarks></remarks>
        public abstract void Update();

        /// <summary>
        /// All subclasses can communicate using messages
        /// </summary>
        public virtual bool HandleMessage(Telegram msg)
        {
            throw new NotImplementedException();
        }

        /// <summary>
        /// Type of the entity
        /// </summary>
        public string Type => this.GetType().ToString();

        /// <summary>
        /// Whether the entity can be used as a camera target
        /// </summary>
        public virtual bool CameraTarget { get; set; } = true;
    }
}
